src/org/traccar/protocol/WatchFrameDecoder.java

```diff
@@ -56,9 +82,20 @@ protected Object decode(
 
         int length = Integer.parseInt(
                 buf.toString(lengthIndex, payloadIndex - lengthIndex, StandardCharsets.US_ASCII), 16);
-        if (buf.readableBytes() >= payloadIndex + 1 + length + 1) {
+
```

Comment:
this part fixes invalid encoded frame length by looking up the next frame end maker ']' and comparing the frame length with the no of bytes till this frame end marker.